### DISCUSSION TITLE
How to execute a command in specific window in a specific workspace

### DISCUSSION THREAD

**Discussion by RayZ0rr** [i3/i3#4816] [Q&A] [answered]
What I have done : 

```sh
i3-msg 'workspace number 4; exec alacritty; exec st'
i3-msg 'workspace number 4; [class="st"] focus; exec --no-startup-id echo foo'
```
But I am not able to get '**foo**' printed out in st terminal. The st terminal however, does get focus (from this `[class="st"] focus`).

**Comment by Airblader** [comment, marked as answer]
i3's `exec` does not and cannot run commands inside terminals. It doesn't even know that some application takes any input, less alone is a terminal.

You can either check your terminal for a command-line option of invoking a command when it is started (`alacritty -e`) or you need to just simulate input with something like `xdotool`.

**Comment by RayZ0rr** [comment]
I managed to get it done with: 
```bash
#!/usr/bin/env bash


i3-msg 'workspace number 1; exec alacritty'
i3-msg 'workspace number 1; exec alacritty'
sleep 1
i3-msg "workspace number 1; [class="Alacritty"] focus; exec --no-startup-id xdotool type $'neofetch\n'"
sleep 1
i3-msg "workspace number 1; [class=Alacritty] focus; exec --no-startup-id xdotool type 'la'"
sleep 1
i3-msg "workspace number 1; [class="Alacritty"] focus; exec --no-startup-id xdotool key Return"
sleep 1
i3-msg 'workspace number 2; exec alacritty -t Vifm'
sleep 1
i3-msg "workspace number 2; [title="Vifm"] focus; exec --no-startup-id xdotool type 'vifm'"
sleep 1
i3-msg "workspace number 2; [title="Vifm"] focus; exec --no-startup-id xdotool key Return"
sleep 1
i3-msg 'workspace number 3; exec alacritty -t Bottom'
sleep 1
i3-msg "workspace number 3; [title="Bottom"] focus; exec --no-startup-id xdotool type 'btm'"
sleep 1
i3-msg "workspace number 3; [title="Bottom"] focus; exec --no-startup-id xdotool key KP_Enter"
sleep 1
i3-msg 'workspace number 1'
```
and in i3 config: 
`exec /path/to/above/file &`
(make it executable first)

The setup I wanted (commands entered are shown in center) : 
In workspace number 1
```
+-Win 1--------+----Win 2----+
|              |            |
|  neofetch    |    la      |
|              |            |
|              |            |
+--------------+------------+
```
In workspace number 2
```
+------Title Vifm---------+
|                        |
|          vifm          |
|                        |
|                        |
+------------------------+
```
In workspace number 3
```
+----Title Bottom-------+
|                        |
|          btm           |
|                        |
|                        |
+------------------------+
```

**Reply by Airblader** [reply]
Using layout saving / restoring and `alacritty -e` would probably be a bit easier, faster, and less error-prone.

**Reply by RayZ0rr** [reply]
@Airblader Can I speed this up anymore? Like with `i3-msg "workspace number 2; exec sleep 10"`? I tried that but some commands were executed before the terminal opened, some were executed simultaneously, and other problems things like that.

Do you have a similar optimized setup to spawn clients at startup?

**Reply by Airblader** [reply]
> Using layout saving / restoring and alacritty -e would probably be a bit easier, faster, and less error-prone.

https://i3wm.org/docs/layout-saving.html

**Reply by RayZ0rr** [reply]
Thanks @Airblader. I can use `alacritty -e` but when using `alacritty -e vifm`, vifm doesn't get other environment variables ( like vim becomes my editor instead of `$EDITOR=nvim` in my bashrc/zshrc).

**Reply by Airblader** [reply]
`alacritty -e vifm` doesn't start an interactive shell. You could do something like `alacritty -e /bin/sh -c "vifm && /bin/bash"` (assuming you use bash), see https://github.com/alacritty/alacritty/issues/1266.
